fabric-recipe-api-v1/src/main/java/net/fabricmc/fabric/impl/recipe/RecipeManagerImpl.java

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.recipe;
+
+import java.util.Map;
+import java.util.function.Function;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.gson.JsonElement;
+import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.recipe.Recipe;
+import net.minecraft.recipe.RecipeManager;
+import net.minecraft.recipe.RecipeType;
+import net.minecraft.util.Identifier;
+
+import net.fabricmc.fabric.api.recipe.v1.RecipeLoadingEvents;
+
+@ApiStatus.Internal
+public final class RecipeManagerImpl {
+	/**
+	 * Stores the static recipes which are added to the {@link net.minecraft.recipe.RecipeManager} when recipes are
+	 * loaded.
+	 */
+	private static final Map<Identifier, Recipe<?>> STATIC_RECIPES = new Object2ObjectOpenHashMap<>();
+	private static final Logger LOGGER = LogManager.getLogger();
+
+	private RecipeManagerImpl() {
+		throw new UnsupportedOperationException("Someone tampered with the universe.");
+	}
+
+	public static void registerStaticRecipe(Recipe<?> recipe) {
+		if (STATIC_RECIPES.put(recipe.getId(), recipe) != null) {
+			throw new IllegalArgumentException("Cannot register " + recipe.getId()
```

Comment:
`Map.put()` can replace the previous entry if it exists, so we need to use something like this:
```java
if (STATIC_RECIPES.contains(recipe)) {
    throw new IllegalArgumentException("Cannot register " + recipe.getId() + " as another recipe with the same identifier already exists.");
} else {
    STATIC_RECIPES.put(recipe.getId(), recipe)
}
```

Reply:
`Map.put()` returns the previous value.

If the previous value is non-null then that means the STATIC_RECIPES already contained the recipe and we throw the error.
The error will crash the game.

One concern might be what happens if a mod catch the exception though.

@liach initially proposed to replace the contains with a null check on put

Reply:
You aren't supposed to catch this illegal argument exception.

If you really hate replacing, you can call `map.putIfAbsent` than making 2 map calls.

Reply:
`putIfAbsent` seems to be the safest while still able to throw the exception. Might change to that then.